### PR TITLE
feat: add cypress-autorecord to mock requests

### LINKED ISSUE #1103: Mock all Cypress requests
**Note:** Please see the discussion below about using `cypress-autorecord` - this is now the preferred approach.

### Overview

For our Cypress tests, we currently mock only the data that is required by the UI elements and flows that we actually test.

We don't, however, mock the remaining calls, which slows down our test runs.

### Reference

See the current mocks for [external requests](https://github.com/shapeshift/web/blob/423b94c43ec0809b3c386cb584ab2dcaeb2a2a00/cypress/support/commands.ts#L75) (those services not managed by ShapeShift), and [internal requests](https://github.com/shapeshift/web/blob/423b94c43ec0809b3c386cb584ab2dcaeb2a2a00/cypress/support/commands.ts#L114) (those that ShapeShift are responsible for) as an example of what we are looking to extend.

To find out what requests Cypress is currently making, we can run Cypress locally using `yarn test:cypress`, which will launch a local Cypress runner. Run the tests with Electron, and monitor the network tab for requests. 

Examples of requests to mock:

- CoinGecko API, e.g. `https://api.coingecko.com/api/*`
- Yearn Finance API
- GitHack

Note, running Cypress locally requires `web` to be running locally - please see the [README](https://github.com/shapeshift/web#readme) for instructions on doing this.

Some of the responses from these requests are large - the bounty hunter should use their discretion by only including data that is actually used by ShapeShift Web. We want to avoid having unnecessarily large mock data files in the codebase, and also avoid the illusion of precision that these mock files might give. The responses to these requests are out of our control, and could change at any time - only mocking the data we use is one way to mitigate the risk of mock data becoming stale.

No-opping requests that we don't need in the test environment is ok.

To keep this work decoupled from the test-related mock data, we probably want to have a new set a new set of Cypress commands, `mockExternalEnvironmentRequests`, and `mockInternalEnvironmentRequests` (or similar - use own discretion). There will likely sit in their own helper file in the `cypress/support` directory.

### Acceptance Criteria

- All Cypress requests that occur whilst running the current test suite are mocked
- Only data that is used by web and Cypress is mocked - i.e. excessively large mock files are avoided
- We continue to use the factory pattern in `cypress/factories` for new mock data
- Code is extensible as we continue to mock additional services in the future, as needed
- Each stub should be given an alias, e.g. `cy.intercept('GET', 'some-url', { responseKey: 'responseValue' }).as('someAlias')`
- The new mocks should be used in existing tests

### Screenshots/Mockups

In the example below, only the request with the alias `getChainLinkData` is mocked - in the resulting PR, we want to mock the rest.

Screen Shot 2022-03-03 at 11 54 21 am

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### PR DESCRIPTION
**Pull request by codinglittlecat**
## Description

- add cypress-autorecord (https://github.com/Nanciee/cypress-autorecord)
## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)
This PR closes #1103

### CODE REVIEW

**Comment by 0xdef1cafe** [comment]
@0xApotheosis structurally looks great to me, if it meets your requirements it gets my stamp

**Comment by 0xApotheosis** [comment]
Thanks @codinglittlecat! I'll take a look at this in the next 24 hours 👌

**Comment by 0xApotheosis** [comment]
@codinglittlecat at first glance, we don't seem to have the mocked responses committed to version control.

Are you able to run `cypress-autorecord` and do this, please?

**Comment by codinglittlecat** [comment]
@0xApotheosis mock datat is generated automatically when you run `yarn cypress:test`

File: cypress/README.md [hunk 1]
```diff
@@ -35,6 +35,24 @@ See best [the docs on best practice for selectors](https://docs.cypress.io/guide
 
 Follow [Cypress best practice](https://docs.cypress.io/guides/references/best-practices)
 
+### How to use cypress-autorecord
+To allow for auto-recording and stubbing to work, require `cypress-autorecord` in each of your test file and call the function at the beginning of your parent describe block.
+
+```js
```
Comment (0xApotheosis) [inline cypress/README.md:41]: As we use Typescript, can you update this to `ts` and update the code style to match our lint? E.g. remove `;`.

File: cypress/README.md [hunk 2]
```diff
@@ -35,6 +35,24 @@ See best [the docs on best practice for selectors](https://docs.cypress.io/guide
...
+const autoRecord = require('cypress-autorecord'); // Require the autorecord function
+  
+describe('Home Page', function() { // Do not use arrow functions
+  autoRecord(); // Call the autoRecord function at the beginning of your describe block
```
Comment (0xApotheosis) [inline cypress/README.md:45]: We don't actually want to call this every time - only initially to get the mocks and commit them to version control.

File: cypress/support/index.ts [hunk 3]
```diff
@@ -1 +1,5 @@
 import './commands'
+Cypress.on('window:before:load', win => {
```
Comment (0xApotheosis) [inline cypress/support/index.ts:2]: Can you please add a comment on what this does/why it is needed?

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
At the moment we are regenerating all of the mocks every time `yarn test:cypress` is run. This is triggered by the `autoRecord()` added to the beginning of our tests.

Instead, in this PR we want to commit the generated tests, but not re-generate them every time. So, we can remove `autoRecord()` from the tests, and update `cypress.json` with:

```ts
"autorecord": {
  "forceRecord": false
}
```

We'll also want to clean our tests after each run by adding:

```ts
"cleanMocks": true
```

And update the README to include instructions on adding mock data to new tests, e.g. using:

```ts
 "recordTests": ["new test"]
```

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
A few more things to fix:

- The "nav bar works" test doesn't seem to be using the mocks
- Running all tests at once generates a new set of mock files under a `All Integration Specs` label - we'll need to commit these mocks to version control, too
- When running the tests with the mocks we get a number of exceptions, causing the tests to fail - this might be something to do with the speed that the tests now run at. We'll need to find a solution for these

**Comment by 0xApotheosis** [comment]
Otherwise, we are looking good @codinglittlecat - the tests do run a lot faster with these mocks in place! Nice work.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Thanks for the updates @codinglittlecat!

This is still not quite right. When I checkout this branch, run `yarn test:cypress`, and run all integration tests, the committed mocks are not used.

It might be the case that we need to use `autoRecord()` in each test, but then disable it in the config, I'm not sure.

I also suspect the mocks we want committed are the ones generated when _all_ tests are run (which I suspect will stored under `cypress/mocks/All Integration Specs.json`) - at the moment it seems like you're generating the mocks by running each set of tests individually - but as the CI pipeline runs them all at once, we'll want to match this behaviour.

In the screenshot below, it shows that the mocks are still being modified after I made some tweaks to get it using them again (perhaps `"forceRecord": false` didn't have the effect we thought it would).

Screen Shot 2022-03-21 at 9 32 03 am

Also, the `nav bar works` test is still not being mocked, even when using the mocks committed under `mocks/dashboard_spec.json`. See screenshot below:

Screen Shot 2022-03-21 at 9 39 58 am
 
In summary:

- Commit the mocks generated when _all_ tests are run, and remove the rest
- Ensure that the PR is in a state such that when it is checked out, and `yarn cypress:run` is run locally, no mocks are modified - it should use the committed mocks
- Ensure we have a _working_ mechanism (ideally a flag) to update mocks in the future, and to clear out stale any mocks
- Ensure all requests are mocked (e.g. those in "nav bar works")

Please request another review when you are happy that these requirements are met, or reach out to me on Discord with any questions 😊

**Comment by 0xApotheosis** [comment]
@codinglittlecat are you able to get these changes to us within 24 hours?

If not we'll need to bring this in-house for the sake of expediency.

**Comment by 0xdef1cafe** [comment]
hey @codinglittlecat - are you also @visitskyworld?

**Comment by 0xApotheosis** [comment]
@codinglittlecat / @visitskyworld I can see a few commits since my last comment, though they only seemed to have addressed one of the remaining 5 outstanding concerns (and most crucially, Cypress is still reaching out to the server for _every_ request - preventing this from occurring is the key requirement that this PR needs to achieve).

Please see below a summary of where I currently see this PR with respect to the requests I've made:

---

✅  Commit the mocks generated when all tests are run, and remove the rest

❌  Ensure that the PR is in a state such that when it is checked out, and `yarn cypress:run` is run locally, no mocks are modified - it should use the committed mocks (currently it's still hitting the server for _every_ request) - i.e., this PR doesn't achieve the primary objective of the bounty

❌  Ensure we have a working mechanism (ideally a flag) to update mocks in the future, and to clear out stale any mocks

❌  Ensure all requests are mocked (e.g. those in "nav bar works")

❌  When running the tests with the mocks we get a number of exceptions, causing the tests to fail - this might be something to do with the speed that the tests now run at. We'll need to find a solution for these

---

As this is dragging on a bit now, we'll take this in house or re-bounty it.

Due to the quality and responsiveness of our interactions on this bounty, and the previous bounty you attempted, we've had to make the difficult call not to accept you for future bounties moving forward - please do not apply.

I appreciate the attempt, and I'd encourage you to continue developing your skills and seeking out other open-source projects to contribute to.
